core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -2136,9 +2137,24 @@ private[spark] class DAGScheduler(
     }
   }
 
-  private[scheduler] def handleShuffleMergeFinalized(stage: ShuffleMapStage): Unit = {
-    stage.shuffleDep.markShuffleMergeFinalized
-    processShuffleMapStageCompletion(stage)
+  private[scheduler] def handleRegisterMergeStatuses(
+      stage: ShuffleMapStage,
+      mergeStatuses: Seq[(Int, MergeStatus)]): Unit = {
+    // Register merge statuses if the stage is still running and shuffle merge is not finalized yet.
+    if (runningStages.contains(stage) && !stage.shuffleDep.shuffleMergeFinalized) {
+      mapOutputTracker.registerMergeResults(stage.shuffleDep.shuffleId, mergeStatuses)
+    }
+  }
+
+  private[scheduler] def handleShuffleMergeFinalized(
+      stage: ShuffleMapStage): Unit = {
+    // Only update MapOutputTracker metadata if the stage is still active. i.e not cancelled.
+    if (runningStages.contains(stage)) {
+      stage.shuffleDep.markShuffleMergeFinalized()
+      processShuffleMapStageCompletion(stage)
+    } else {
+      mapOutputTracker.unregisterAllMergeResult(stage.shuffleDep.shuffleId)
```

Comment:
If the stage is not part of runningStages does it really mean that it was cancelled? 
It seems like runningStages contain active stages. If a stage is completed successfully, it will not be part of runningStages. Why do we unregister all the mergeResults here ? 
Also please add a comment that why we need to unregister all the merge results here

Reply:
Stage will be active until the shuffle merge is finalized and then only we are processing the map stage completion, isn't it? So if the stage is not part of running stages and we still reach the handling shuffle merge finalize, then we need to unregister the merge results, isn't it?

Can you think of a scenario where stage is not part of running stages and still shuffle merge is finalized? - Ideally this should not happen.

Reply:
> Stage will be active until the shuffle merge is finalized and then only we are processing the map stage completion, isn't it? So if the stage is not part of running stages and we still reach the handling shuffle merge finalize, then we need to unregister the merge results, isn't it?

Are you just making an assumption here, that if the stage is not running and then this finalized message is processed that means the stage is cancelled?  Is this a valid assumption?
If yes, then can you add a comment here. 

> Can you think of a scenario where stage is not part of running stages and still shuffle merge is finalized? - Ideally this should not happen.

This is what is throwing me off. If this is not ideally going to happen then why are we unregistering the results here. Again, if the assumption is that this happens when the stage was cancelled then document it. Also, if handling stage cancellation wrt merge finalization is not handled in this PR then why have this unregistration of merge results here?

Reply:
In scheduler, if a stage is in`runningStages` is the typical way in which we check if a stage is completed or not.
Outside of scheduler, stage.latestInfo.failureReason is used to identify stage failure.

Having said that, latestInfo is overwritten when a new attempt starts.

For this specific case we had discussed (stage resubmission after a cancellation after initiating merge finalization), given a new stage attempt would result in the stage continuing to be in `runningStages` (but for newer attempt), the way to check would be:

a) Fetch `stage.latestInfo.attemptNumber()` in start of `scheduleShuffleMergeFinalize` (within dag scheduler event loop) - and pass this to finalizeShuffleMerge -> RegisterMergeStatuses and ShuffleMergeFinalized.

In handleShuffleMergeFinalized:
b) If stage is not in runningStages (like currently being done), then not running - or
c) If stage.latestInfo.attemptNumber() != ShuffleMergeFinalized.attemptNumber then attempt has changed, and finalization message.

Reply:
On further thought ... my comment about attemptNumber is not valid.
A new job will result in a new stage (since `stage` would have been removed from `shuffleIdToMapStage` and `runningStages` due to cancellation).

I think the code looks fine as is.

+CC @otterc ... thoughts ?

Reply:
@mridulm For the case that you pointed out
> stage resubmission after a cancellation after initiating merge finalization

The problem that we were registering the mergeResults without a check and that would interfere with resubmission is solved here by not registering merge results if the stage is not running. Also, it makes sense to finalize the stage when it is running.

However, shouldn't unregistering of merge results happen when there is a fetch failure or a stage is cancelled?

Both of the above are not part of this PR. So, I just want to understand in the context of this PR, why are we unregistering the merge results in the else part here and what does that help with?
 ```
 else {      mapOutputTracker.unregisterAllMergeResult(stage.shuffleDep.shuffleId)
 ```
 Also if there is a good reason to do it, then we should document it here as a comment.

Reply:
Discussed offline with @mridulm and currently there are few corner cases which needs to be carefully thought through before having this behavior. Created a TODO and a corresponding follow up JIRA - https://issues.apache.org/jira/browse/SPARK-35549